packages/contracts/src/EmailWalletCore.sol

```diff
@@ -995,6 +1050,7 @@ contract EmailWalletCore is ReentrancyGuard, OwnableUpgradeable, UUPSUpgradeable
                 Wallet wallet = Wallet(payable(currContext.walletAddr));
 
                 try wallet.execute(weth, 0, abi.encodeWithSignature("withdraw(uint256)", walletParams.amount)) {
+                    // Sora's comment: Is this `execute` doing raw ETH transfer? 
```

Comment:
Yes, we are converting WETH in the wallet to ETH and then sending to the recipient. 
Wallet contract auto convert all ETH deposits to WETH.

This is based on our discord discussions. However I am not fully happy with all this as there is many moving parts. If we can handle ETH/WETH in a different way, it would be better. I will also think.

Reply:
> This is based on our discord discussions. However I am not fully happy with all this as there is many moving parts. If we can handle ETH/WETH in a different way, it would be better. I will also think.

I think we should not require the user to distinguish between raw ETH and WETH.
Even if they are internally managed separately, the balance of ETH shown to users/extensions is equal to the sum of raw ETH and WETH.
In that case, how exactly will you simplify the implementation?

Reply:
> I think we should not require the user to distinguish between raw ETH and WETH.
I also think so, this would be right for most users. But on the other hand, crypto native users might not like we abstracting this and auto converting ETH/WETH - i.e if someone send them ETH, it auto converts to WETH now.

One other idea is to keep both separate. An average user will only have ETH received from others. And for non-wallet commands, we can let extensions switch between ETH/WETH. 
i.e Wallet dont do any auto conversion
- When user "send 10 eth to someone" real eth gets transferred. If there is only 9eth and 1weth we can probbaly convert and send.
- If user say "send 10 weth to someone" we only send weth and fail if there is no balance (even if there was balance in eth)
- when user "sway swap 10 eth to dai" the uniswap extension can do the eth to weth conversion and then 

This is basically like eoa, except we do auto weht to eth only on transfer. 
This is not simpler proabbly, but we don't do auto conversion in wallet `fallback()` which may not receive enough gas needed for the conversion.

Reply:
Thank you for the detail of your idea.
I understand the key points are 1) "eth" in the subject includes weth but 2) "weth" is distinguished from eth.
However, I think it exposes much complexity to users and extensions.
For example, if the extension developer does not understand this spec and requires "weth" in the subject, its users will be confused because the extension returns an error of insufficient balance.
I think auto-convert in the current spec is better than separating eth and weth.

Reply:
> I think auto-convert in the current spec is better than separating eth and weth.
Yes, simpler for our implementation. The issue is we are converting "on receive" rather than converting "when needed". When the user exit email wallet to own EOA, they might not want this functionality. They will be forced to upgrade the contract (or we disable this auto convert on exits)

Reply:
> or we disable this auto convert on exits
If possible, this is the best.

Reply:
> If possible, this is the best.

Should be possible. But we need to confirm our upgrade plan. If the idea is for wallet to exit to new core contract address, then we should not remove this auto-conversion.

Reply:
> If the idea is for wallet to exit to new core contract address, then we should not remove this auto-conversion.

Oh yes, I missed that point.
Even if we implement the admin upgrade of the core contract for a short term, it should be possible to change the owner to a new core contract by the user's exit command.
Therefore, I think we should not remove the auto-conversion.

Reply:
True. I think the best is to add a function in wallet to disable this functionality, so the new EOA owner can call this to stop auto conversion.

Reply:
The core contract needs to enable the nullifier for the exited wallet.
Otherwise, a malicious wallet sets the owner to the core contract again after disabling the auto-convert functionality (or adding/removing any features that can be DoS attack vectors against a relayer).

Reply:
Hmm, it is still possible for user to exist to EOA, upgrade the wallet contract and set owner back to core. Our wallet contract is upgradeable.
Btw, did you already think of a concrete attack vector? I think Relayer running simulation should prevent it unless user time the upgrade+owner change - (but doing upgrade+owner change has gas cost for user)

Reply:
> Hmm, it is still possible for user to exist to EOA, upgrade the wallet contract and set owner back to core. Our wallet contract is upgradeable.

As long as no relayer can use that wallet, it is fine.
I found the nullifier is useless here because it is stored for each account key commitment.
A more direct way is to store the exited wallet addresses.

> Btw, did you already think of a concrete attack vector? 
For example, 
1) a wallet exits the core contract.
2) the wallet adds an allowance of the fee token (ETH/DAI/USDC) to an extension contract.
3) the wallet goes back to the core contract.
4) the wallet (user) calls that extension.
5) the extension pulls the user's fee tokens by `transferFrom` without writing the token requests in the subject.
6) after the execution, the user does not have a sufficient balance to pay the fee.

The relayer cannot prevent/detect such an attack without application-specific solutions.

Reply:
For the same reason, when the command is `execute`, the `_computeMaskedSubjectForEmailOp` function should check `bytes(tokenRegistry.getTokenNameOfAddress(target)).length == 0`.